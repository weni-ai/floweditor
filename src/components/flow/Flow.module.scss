@import 'variables.module.scss';

.new_updates {

  > div {
    padding: $unnnic-spacing-stack-lg $unnnic-spacing-stack-md !important;
  }

  .content {
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 20px;

    .title {
      color: $unnnic-color-neutral-darkest;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-title-sm;
      font-weight: $unnnic-font-weight-black;
      line-height: $unnnic-line-height-md + $unnnic-font-size-title-sm;
    }

    .description {
      text-align: center;
      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-line-height-md + $unnnic-font-size-body-lg;
    }


    .news_list {
      display: flex;
      gap: $unnnic-spacing-xs;
      flex-direction: column;
      align-items: center;
      text-align: center;
      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-line-height-md + $unnnic-font-size-body-lg;
    }

    .footer {
      color: $unnnic-color-neutral-black;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-line-height-md + $unnnic-font-size-body-lg;
    }
  }

  .buttons {
    margin-top: $unnnic-spacing-stack-lg;
    display: flex;
    gap: $unnnic-spacing-stack-md;
    flex: 1;

    > * {
      flex: 1;

      button {
        width: 100%;
      }
    }
    
    .primary {
      background-color: $unnnic-color-brand-weni-soft;
    }

  }
}