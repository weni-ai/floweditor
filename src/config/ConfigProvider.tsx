import { FlowEditorConfig } from 'flowTypes';
import * as React from 'react';

export const fakePropType: any = (): any => null;
fakePropType.isRequired = (): any => null;

interface ConfigProviderProps {
  config: FlowEditorConfig;
  children: any;
}

export interface ConfigProviderContext {
  config: FlowEditorConfig;
}

// ----------------------------------------------------------------------------------------------

export const SINGLE_CHILD_ERROR =
  'ConfigProvider expects only one child component.';
export const VALID_CHILD_ERROR =
  'ConfigProvider expects a valid React element: https://reactjs.org/docs/react-api.html#isvalidelement';

export default class ConfigProvider extends React.Component<
  ConfigProviderProps
> {
  public static childContextTypes = {
    config: fakePropType,
    assetService: fakePropType,
  };

  constructor(props: ConfigProviderProps) {
    super(props);

    if (React.Children.count(props.children) > 1) {
      throw new Error(SINGLE_CHILD_ERROR);
    } else if (!React.isValidElement(props.children)) {
      throw new Error(VALID_CHILD_ERROR);
    }
  }

  public getChildContext(): ConfigProviderContext {
    return {
      config: this.props.config,
    };
  }

  public render(): JSX.Element {
    return React.Children.only(this.props.children);
  }
}
