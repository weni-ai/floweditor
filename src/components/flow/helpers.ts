import { Type, Types } from 'config/interfaces';
import { getType, getTypeConfig } from 'config/typeConfigs';
import { NodeEditorSettings } from 'store/nodeEditor';
import { Asset } from 'store/flowContext';
import { FlowIssue } from 'flowTypes';

export const determineTypeConfig = (nodeSettings: NodeEditorSettings): Type => {
  const renderNode = nodeSettings.originalNode;
  const node = renderNode && renderNode.node;

  if (nodeSettings.originalAction && nodeSettings.originalAction.type) {
    return getTypeConfig(nodeSettings.originalAction.type);
  } else if (node && node.actions && node.actions.length > 0) {
    return getTypeConfig(node.actions[node.actions.length - 1].type);
  } else {
    try {
      const type = getType(renderNode);
      const config = getTypeConfig(type);
      if (config.type !== Types.missing) {
        return config;
      }
      // tslint:disable-next-line:no-empty
    } catch (Error) {}
  }

  // Account for ghost nodes
  if (node && node.router) {
    return getTypeConfig(node.router.type);
  }

  throw new Error(`Couldn't determine type config for: ${node.uuid}`);
};

export const hasIssues = (
  issues: FlowIssue[],
  translating: boolean,
  language: Asset,
): boolean => {
  if (!issues || issues.length === 0) {
    return false;
  }

  if (!translating) {
    return !!issues.find((issue: FlowIssue) => !issue.language);
  } else {
    return !!issues.find((issue: FlowIssue) => issue.language === language.id);
  }
};
