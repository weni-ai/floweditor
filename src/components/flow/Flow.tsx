import ReactDOM from 'react-dom';
import { react as bindCallbacks } from 'auto-bind';
import { Canvas } from 'components/canvas/Canvas';
import { CanvasDraggableProps } from 'components/canvas/CanvasDraggable';
import Node from 'components/flow/node/Node';
import { getDraggedFrom } from 'components/helpers';
import NodeEditor from 'components/nodeeditor/NodeEditor';
import Simulator from 'components/simulator/Simulator';
import Sidebar from 'components/sidebar/Sidebar';
import Sticky, { STICKY_BODY, STICKY_TITLE } from 'components/sticky/Sticky';
import { ConfigProviderContext, fakePropType } from 'config/ConfigProvider';
import { FlowMetadata, FlowPosition, StickyNote } from 'flowTypes';
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Plumber from 'services/Plumber';
import { DragSelection, DebugState } from 'store/editor';
import { RenderNode } from 'store/flowContext';
import { detectLoops, getOrderedNodes } from 'store/helpers';
import { NodeEditorSettings } from 'store/nodeEditor';
import AppState from 'store/state';
import {
  ConnectionEvent,
  DispatchWithState,
  mergeEditorState,
  MergeEditorState,
  NoParamsAC,
  onConnectionDrag,
  OnConnectionDrag,
  OnOpenNodeEditor,
  onOpenNodeEditor,
  onRemoveNodes,
  OnRemoveNodes,
  OnUpdateCanvasPositions,
  onUpdateCanvasPositions,
  resetNodeEditingState,
  UpdateConnection,
  updateConnection,
  UpdateNodesEditor,
  updateNodesEditor,
  updateSticky,
  UpdateSticky
} from 'store/thunks';
import {
  createUUID,
  isRealValue,
  NODE_PADDING,
  renderIf,
  snapToGrid,
  timeEnd,
  timeStart
} from 'utils';
import Debug from 'utils/debug';

import { PopTabType } from 'config/interfaces';

import styles from './Flow.module.scss';
import { applyVueInReact } from 'vuereact-combined';
// @ts-ignore
import { unnnicModalNext, unnnicButton } from '@weni/unnnic-system';
import { WeniLoveIcon } from './WeniLoveIcon';
import i18n from '../../config/i18n';

const UnnnicButton = applyVueInReact(unnnicButton, {
  vue: {
    componentWrap: 'div',
    slotWrap: 'div',
    componentWrapAttrs: {
      style: {
        all: ''
      }
    }
  }
});

const UnnnicModalNext = applyVueInReact(unnnicModalNext, {
  vue: {
    componentWrap: 'div',
    slotWrap: 'div',
    componentWrapAttrs: {
      style: {
        all: '',
        position: 'relative',
        zIndex: 10e2
      }
    }
  }
});

declare global {
  interface Window {
    fe: any;
  }
}
export interface FlowStoreProps {
  ghostNode: RenderNode;
  debug: DebugState;
  translating: boolean;
  popped: string;
  dragActive: boolean;

  mergeEditorState: MergeEditorState;

  stickyMap: { [key: string]: StickyNote };
  nodes: { [uuid: string]: RenderNode };
  metadata: FlowMetadata;
  nodeEditorSettings: NodeEditorSettings;

  updateConnection: UpdateConnection;
  onOpenNodeEditor: OnOpenNodeEditor;
  onUpdateCanvasPositions: OnUpdateCanvasPositions;
  onRemoveNodes: OnRemoveNodes;
  resetNodeEditingState: NoParamsAC;
  onConnectionDrag: OnConnectionDrag;
  updateSticky: UpdateSticky;

  updateNodesEditor?: UpdateNodesEditor;
}

export interface Translations {
  [uuid: string]: any;
}

export const DRAG_THRESHOLD = 3;
export const REPAINT_TIMEOUT = 500;
export const GHOST_POSITION_INITIAL = { left: -1000, top: -1000 };

export const nodeSpecId = 'node';
export const nodesContainerSpecId = 'node-container';
export const ghostNodeSpecId = 'ghost-node';
export const dragSelectSpecId = 'drag-select';

export const isDraggingBack = (event: ConnectionEvent) => {
  return event.suspendedElementId === event.targetId && event.source !== null;
};

export const getDragStyle = (drag: DragSelection) => {
  const left = Math.min(drag.startX, drag.currentX);
  const top = Math.min(drag.startY, drag.currentY);
  const width = Math.max(drag.startX, drag.currentX) - left;
  const height = Math.max(drag.startY, drag.currentY) - top;
  return {
    left,
    top,
    width,
    height
  };
};

export class Flow extends React.PureComponent<FlowStoreProps, {}> {
  private Plumber: Plumber;
  private nodeContainerUUID: string;
  private canvas: React.RefObject<Canvas>;

  // Refs
  private ghost: any;

  public static contextTypes = {
    config: fakePropType
  };

  constructor(props: FlowStoreProps, context: ConfigProviderContext) {
    super(props, context);

    this.canvas = React.createRef();

    this.nodeContainerUUID = createUUID();

    this.Plumber = new Plumber();

    /* istanbul ignore next */
    if (context.config.debug) {
      window.fe = new Debug(props, this.props.debug);
    }

    bindCallbacks(this, {
      include: [/Ref$/, /^on/, /^is/, /^get/, /^handle/]
    });

    timeStart('Loaded Flow');
  }

  private ghostRef(ref: any): any {
    return (this.ghost = ref);
  }

  public componentDidMount(): void {
    this.Plumber.bind('connection', (event: ConnectionEvent) =>
      this.props.updateConnection(event.sourceId, event.targetId)
    );
    this.Plumber.bind('beforeDrag', (event: ConnectionEvent) => {
      this.beforeConnectionDrag(event);
    });

    this.Plumber.bind('connectionDrag', (event: ConnectionEvent) => {
      this.props.onConnectionDrag(event, this.context.config.flowType);
    });

    this.Plumber.bind('connectionDragStop', (event: ConnectionEvent) =>
      this.onConnectorDrop(event)
    );
    this.Plumber.bind(
      'beforeStartDetach',
      (event: ConnectionEvent) => !this.props.translating && this.context.config.mutable
    );
    this.Plumber.bind('beforeDetach', (event: ConnectionEvent) => true);
    this.Plumber.bind('beforeDrop', (event: ConnectionEvent) => this.onBeforeConnectorDrop(event));
    this.Plumber.triggerLoaded(this.context.config.onLoad);

    this.checkToShowNewUpdates();

    timeEnd('Loaded Flow');
  }

  public componentWillUnmount(): void {
    this.Plumber.reset();
  }

  /**
   * Called right before a connector is dropped onto a new node
   */
  private onBeforeConnectorDrop(event: ConnectionEvent): boolean {
    this.props.resetNodeEditingState();
    const fromNodeUUID = event.sourceId.split(':')[0];
    try {
      detectLoops(this.props.nodes, fromNodeUUID, event.targetId);
    } catch {
      return false;
    }
    return true;
  }

  /**
   * Called the moment a connector is done dragging, whether it is dropped on an
   * existing node or on to empty space.
   */
  private onConnectorDrop(event: ConnectionEvent): boolean {
    const ghostNode = this.props.ghostNode;
    // Don't show the node editor if we a dragging back to where we were
    if (isRealValue(ghostNode) && !isDraggingBack(event)) {
      // Wire up the drag from to our ghost node
      this.Plumber.recalculate(ghostNode.node.uuid);

      const dragPoint = getDraggedFrom(ghostNode);

      this.Plumber.connect(dragPoint.nodeUUID + ':' + dragPoint.exitUUID, ghostNode.node.uuid);

      // Save our position for later
      const { left, top } = (this.ghost &&
        snapToGrid(this.ghost.ele.offsetLeft, this.ghost.ele.offsetTop)) || { left: 0, top: 0 };

      this.props.ghostNode.ui.position = { left, top };

      let originalAction = null;
      if (ghostNode.node.actions && ghostNode.node.actions.length === 1) {
        originalAction = ghostNode.node.actions[0];
      }

      // Bring up the node editor
      this.props.onOpenNodeEditor({
        originalNode: ghostNode,
        originalAction
      });
    }

    if (isDraggingBack(event)) {
      this.props.mergeEditorState({ ghostNode: null });
    }

    /* istanbul ignore next */
    document.removeEventListener('mousemove', (window as any).ghostListener);

    return true;
  }

  private beforeConnectionDrag(event: ConnectionEvent): boolean {
    if (event.source) {
      event.source.dispatchEvent(new Event('disconnect'));
    }
    return !this.props.translating;
  }

  private handleStickyCreation(props: CanvasDraggableProps) {
    const stickyMap = this.props.stickyMap || {};
    const uuid = props.uuid;
    return <Sticky key={uuid} uuid={uuid} sticky={stickyMap[uuid]} selected={props.selected} />;
  }

  private handleNodeCreation(props: CanvasDraggableProps) {
    const onlyNode = Object.keys(this.props.nodes).length === 1;
    return (
      <Node
        onlyNode={onlyNode}
        startingNode={props.idx === 0}
        selected={props.selected}
        key={props.uuid}
        data-spec={nodeSpecId}
        nodeUUID={props.uuid}
        plumberMakeTarget={this.Plumber.makeTarget}
        plumberRemove={this.Plumber.remove}
        plumberRecalculate={this.Plumber.recalculate}
        plumberMakeSource={this.Plumber.makeSource}
        plumberConnectExit={this.Plumber.connectExit}
        plumberUpdateClass={this.Plumber.updateClass}
      />
    );
  }

  private getNodes(): CanvasDraggableProps[] {
    return getOrderedNodes(this.props.nodes).map((renderNode: RenderNode, idx: number) => {
      return {
        uuid: renderNode.node.uuid,
        position: renderNode.ui.position,
        elementCreator: this.handleNodeCreation,
        config: renderNode,
        idx
      };
    });
  }

  private getStickies(): CanvasDraggableProps[] {
    const stickyMap = this.props.stickyMap || {};
    return Object.keys(stickyMap).map((uuid: string, idx: number) => {
      return {
        uuid,
        elementCreator: this.handleStickyCreation,
        position: stickyMap[uuid].position,
        idx
      };
    });
  }

  private getDragNode(): JSX.Element {
    return isRealValue(this.props.ghostNode) ? (
      <div
        data-spec={ghostNodeSpecId}
        key={this.props.ghostNode.node.uuid}
        style={{ position: 'absolute', display: 'block', visibility: 'hidden' }}
      >
        <Node
          onlyNode={false}
          selected={false}
          startingNode={false}
          ref={this.ghostRef}
          ghost={true}
          nodeUUID={this.props.ghostNode.node.uuid}
          plumberMakeTarget={this.Plumber.makeTarget}
          plumberRemove={this.Plumber.remove}
          plumberRecalculate={this.Plumber.recalculate}
          plumberMakeSource={this.Plumber.makeSource}
          plumberConnectExit={this.Plumber.connectExit}
          plumberUpdateClass={this.Plumber.updateClass}
        />
      </div>
    ) : null;
  }

  private getSimulator(): JSX.Element {
    return renderIf(this.context.config.endpoints && this.context.config.endpoints.simulateStart)(
      <Simulator
        key="simulator"
        popped={this.props.popped}
        mergeEditorState={this.props.mergeEditorState}
        onToggled={(visible: boolean, tab: PopTabType) => {
          this.props.mergeEditorState({
            popped: visible ? tab : null
          });
        }}
      />
    );
  }

  private getNodeEditor(): JSX.Element {
    return renderIf(this.props.nodeEditorSettings !== null)(
      <NodeEditor
        key="node-editor"
        helpArticles={this.context.config.help}
        plumberConnectExit={this.Plumber.connectExit}
      />
    );
  }

  // TODO: this should be a callback from the canvas
  private handleDoubleClick(position: FlowPosition): void {
    const { left, top } = position;
    this.props.updateSticky(createUUID(), {
      position: snapToGrid(left - 90 + NODE_PADDING, top - 40),
      title: STICKY_TITLE,
      body: STICKY_BODY
    });
  }

  public handleDragging(uuids: string[]): void {
    uuids.forEach((uuid: string) => {
      try {
        const ele = document.getElementById(uuid);
        const exits = ele.querySelectorAll('.jtk-connected');
        this.Plumber.revalidate([ele, ...exits]);
      } catch (error) {}
    });
  }

  public handleCanvasLoaded(): void {
    this.Plumber.setContainer('canvas');
  }

  private callCanvasCopy(): void {
    this.canvas.current.manuallyCopy();
  }

  private startGuiding() {
    this.props.mergeEditorState({ currentGuide: 'v2', guidingStep: 0 });
  }

  private showNewUpdatesModal(): void {
    let newUpdatesModalEl: HTMLDivElement = document.querySelector('#new-updates-modal');

    if (!newUpdatesModalEl) {
      newUpdatesModalEl = document.createElement('div');
      newUpdatesModalEl.setAttribute('id', 'new-updates-modal');
      document.body.appendChild(newUpdatesModalEl);
    }

    if (newUpdatesModalEl.hasChildNodes()) {
      return;
    }

    ReactDOM.render(
      <UnnnicModalNext className={styles.new_updates}>
        <div className={styles.content}>
          <WeniLoveIcon />
          <span className={styles.title}>
            {i18n.t('new_updates.title', 'Flow Editor has a new look!')}
          </span>
          <span className={styles.description}>
            {i18n.t('new_updates.description', 'Whats new?')}
          </span>

          <div className={styles.news_list}>
            <span>•&nbsp; {i18n.t('new_updates.news.first', 'A beautiful new interface')}</span>
            <span>
              •&nbsp; {i18n.t('new_updates.news.second', 'Nodes copy and paste functionality')}
            </span>
            <span>•&nbsp; {i18n.t('new_updates.news.third', 'Sidebar with tools (evolving)')}</span>
            <span>
              •&nbsp;{' '}
              {i18n.t('new_updates.news.fourth', 'Possibility to create a new block anywhere')}
            </span>
            <span>•&nbsp; {i18n.t('new_updates.news.fifth', 'WhatsApp simulator skin')}</span>
          </div>
        </div>

        <div className={styles.buttons}>
          <UnnnicButton
            type="ghost"
            text={i18n.t('new_updates.buttons.first', 'Ignore')}
            onClick={() => {
              ReactDOM.unmountComponentAtNode(newUpdatesModalEl);
            }}
          />
          <UnnnicButton
            className={styles.primary}
            type="primary"
            text={i18n.t('new_updates.buttons.second', 'Show me everything')}
            onClick={() => {
              this.startGuiding();
              ReactDOM.unmountComponentAtNode(newUpdatesModalEl);
            }}
          />
        </div>
      </UnnnicModalNext>,
      newUpdatesModalEl
    );
  }

  private checkToShowNewUpdates() {
    if (this.context.config.showNewUpdates) {
      this.showNewUpdatesModal();
    }
  }

  public render(): JSX.Element {
    const nodes = this.getNodes();

    const draggables = this.getStickies().concat(nodes);

    return (
      <div>
        {this.getSimulator()}
        {this.getNodeEditor()}

        <Sidebar onCopyClick={() => this.callCanvasCopy()} />

        <Canvas
          ref={this.canvas}
          mutable={this.context.config.mutable}
          draggingNew={!!this.props.ghostNode && !this.props.nodeEditorSettings}
          newDragElement={this.getDragNode()}
          onDragging={this.handleDragging}
          uuid={this.nodeContainerUUID}
          dragActive={this.props.dragActive}
          mergeEditorState={this.props.mergeEditorState}
          onRemoveNodes={this.props.onRemoveNodes}
          draggables={draggables}
          onDoubleClick={this.handleDoubleClick}
          onUpdatePositions={this.props.onUpdateCanvasPositions}
          onLoaded={this.handleCanvasLoaded}
          nodes={this.props.nodes}
          updateNodesEditor={this.props.updateNodesEditor}
        ></Canvas>
        <div id="activity_recent_messages"></div>
      </div>
    );
  }
}

/* istanbul ignore next */
const mapStateToProps = ({
  flowContext: { definition, metadata, nodes },
  editorState: { ghostNode, debug, translating, popped, dragActive },
  // tslint:disable-next-line: no-shadowed-variable
  nodeEditor: { settings }
}: AppState) => {
  return {
    nodeEditorSettings: settings,
    stickyMap: definition._ui.stickies,
    nodes,
    metadata,
    ghostNode,
    debug,
    translating,
    popped,
    dragActive
  };
};

/* istanbul ignore next */
const mapDispatchToProps = (dispatch: DispatchWithState) =>
  bindActionCreators(
    {
      mergeEditorState,
      resetNodeEditingState,
      onConnectionDrag,
      onOpenNodeEditor,
      onUpdateCanvasPositions,
      onRemoveNodes,
      updateConnection,
      updateSticky,
      updateNodesEditor
    },
    dispatch
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Flow);
