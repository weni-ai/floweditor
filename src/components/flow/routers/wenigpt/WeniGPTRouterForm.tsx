import { react as bindCallbacks } from 'auto-bind';
import Dialog, { ButtonSet } from 'components/dialog/Dialog';
import { renderIssues } from 'components/flow/actions/helpers';
import { RouterFormProps } from 'components/flow/props';
import { nodeToState, stateToNode } from 'components/flow/routers/wenigpt/helpers';
import { createResultNameInput } from 'components/flow/routers/widgets';
import SelectElement from 'components/form/select/SelectElement';
import TextInputElement from 'components/form/textinput/TextInputElement';
import TypeList from 'components/nodeeditor/TypeList';
import * as React from 'react';
import { FormEntry, FormState, mergeForm, StringEntry } from 'store/nodeEditor';
import { shouldRequireIf, validate } from 'store/validators';

import styles from './WeniGPTRouterForm.module.scss';
import i18n from 'config/i18n';

export interface WeniGPTRouterFormState extends FormState {
  knowledgeBase: FormEntry;
  expression: StringEntry;
  resultName: StringEntry;
  knowledgeBases: any[];
}

export default class WeniGPTRouterForm extends React.Component<
  RouterFormProps,
  WeniGPTRouterFormState
> {
  constructor(props: RouterFormProps) {
    super(props);
    this.state = nodeToState(this.props.nodeSettings, this.props.assetStore);
    bindCallbacks(this, {
      include: [/^handle/]
    });
  }

  private handleUpdate(
    keys: {
      knowledgeBase?: any;
      expression?: string;
      resultName?: string;
    },
    submitting = false
  ): boolean {
    const updates: Partial<WeniGPTRouterFormState> = {};

    if (keys.hasOwnProperty('knowledgeBase')) {
      updates.knowledgeBase = validate(
        i18n.t('forms.knowledge_base', 'Knowledge Base'),
        keys.knowledgeBase,
        [shouldRequireIf(submitting)]
      );
    }

    if (keys.hasOwnProperty('expression')) {
      updates.expression = validate(i18n.t('forms.expression', 'Expression'), keys.expression, [
        shouldRequireIf(submitting)
      ]);
    }

    if (keys.hasOwnProperty('resultName')) {
      updates.resultName = validate(i18n.t('forms.result_name', 'Result Name'), keys.resultName, [
        shouldRequireIf(submitting)
      ]);
    }

    const updated = mergeForm(this.state, updates);

    // update our form
    this.setState(updated);
    return updated.valid;
  }

  private handleUpdateResultName(value: string): boolean {
    return this.handleUpdate({ resultName: value });
  }

  private handleExpressionUpdate(expression: string): boolean {
    return this.handleUpdate({ expression });
  }

  private handleKnowledgeBaseUpdate(knowledgeBase: any): boolean {
    return this.handleUpdate({ knowledgeBase });
  }

  private handleRedirectClick(): void {
    window.parent.postMessage({ event: 'redirect', path: 'intelligences:init/force' }, '*');
  }

  private handleSave(): void {
    const valid = this.handleUpdate(
      {
        knowledgeBase: this.state.knowledgeBase.value,
        expression: this.state.expression.value,
        resultName: this.state.resultName.value
      },
      true
    );

    if (valid) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.save'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  private renderEdit(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    return (
      <Dialog title={typeConfig.name} headerClass={typeConfig.type} buttons={this.getButtons()}>
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <div className={styles.content}>
          <div className={styles.knowledge_base}>
            <SelectElement
              key="knowledge_base_select"
              name={i18n.t('forms.knowledge_base', 'Knowledge Base')}
              placeholder={i18n.t('forms.knowledge_base_placeholder', 'Select a knowledge base')}
              showLabel={true}
              entry={this.state.knowledgeBase}
              onChange={this.handleKnowledgeBaseUpdate}
              options={this.state.knowledgeBases}
              nameKey="label"
              valueKey="id"
            />

            <div className={styles.message}>
              <span>
                {i18n.t(
                  'forms.knowledge_base_message',
                  "Don't have any knowledge base for your AI yet?"
                )}
              </span>
              <span className={styles.link} onClick={this.handleRedirectClick}>
                {i18n.t('forms.knowledge_base_click', 'Click here')}
              </span>
            </div>
          </div>

          <TextInputElement
            name={i18n.t('forms.expression_input', 'Insert an expression to be used as input')}
            showLabel={true}
            placeholder={i18n.t('forms.expression_input_placeholder', 'Ex: @input.text')}
            entry={this.state.expression}
            onChange={this.handleExpressionUpdate}
            autocomplete={true}
            textarea={true}
          />
        </div>
        {createResultNameInput(this.state.resultName, this.handleUpdateResultName)}
        {renderIssues(this.props)}
      </Dialog>
    );
  }

  public render(): JSX.Element {
    return this.renderEdit();
  }
}
