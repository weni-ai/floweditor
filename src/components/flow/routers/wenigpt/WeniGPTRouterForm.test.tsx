/* eslint-disable @typescript-eslint/no-object-literal-type-assertion */
import { RouterFormProps } from 'components/flow/props';
import WeniGPTRouterForm from 'components/flow/routers/wenigpt/WeniGPTRouterForm';
import { Types } from 'config/interfaces';
import { AssetType, RenderNode } from 'store/flowContext';
import { composeComponentTestUtils, mock } from 'testUtils';
import { createWebhookRouterNode, getRouterFormProps } from 'testUtils/assetCreators';
import * as utils from 'utils';
import * as React from 'react';
import {
  render,
  fireEvent,
  fireUnnnicInputChangeText,
  fireUnnnicTextAreaChangeText,
  act
} from 'test/utils';
import userEvent from '@testing-library/user-event';
let mockedKnowledgeBases = require('test/assets/knowledge_bases.json');
mockedKnowledgeBases.forEach((kb: any) => {
  kb.content = kb;
});

mock(utils, 'createUUID', utils.seededUUIDs());

const weniGPTForm = getRouterFormProps({
  node: createWebhookRouterNode(),
  ui: { type: Types.call_wenigpt }
} as RenderNode);

weniGPTForm.assetStore.knowledgeBases = {
  type: AssetType.KnowledgeBase,
  items: mockedKnowledgeBases
};

const { setup } = composeComponentTestUtils<RouterFormProps>(WeniGPTRouterForm, weniGPTForm);

describe(WeniGPTRouterForm.name, () => {
  it('should render', () => {
    const { wrapper } = setup(true);
    expect(wrapper).toMatchSnapshot();
  });

  describe('updates', () => {
    it('should save changes', async () => {
      const { getByText, getByTestId } = render(<WeniGPTRouterForm {...weniGPTForm} />);

      const okButton = getByText('Save');

      // our fields are required
      fireEvent.click(okButton);
      expect(weniGPTForm.updateRouter).not.toBeCalled();

      // set our result
      const resultName = getByTestId('Save as result');
      await act(async () => {
        fireUnnnicInputChangeText(resultName, 'My Webhook Result');
      });

      // set our knowledge_base
      userEvent.click(getByText('Intelligence 2 - Base 1'));

      // set our expression input
      const expressionInput = getByTestId('Insert an expression to be used as input');
      await act(async () => {
        fireUnnnicTextAreaChangeText(expressionInput, 'Updated expression input data');
      });

      userEvent.click(okButton);
      expect(weniGPTForm.updateRouter).toBeCalled();
      expect(weniGPTForm.updateRouter).toMatchCallSnapshot();
    });
  });

  it('should cancel', () => {
    const { instance, props } = setup(true, {
      $merge: { onClose: jest.fn(), updateRouter: jest.fn() }
    });
    instance.getButtons().secondary.onClick();
    expect(props.onClose).toHaveBeenCalled();
    expect(props.updateRouter).not.toHaveBeenCalled();
  });

  it('should redirect', () => {
    const { getByText } = render(<WeniGPTRouterForm {...weniGPTForm} />);

    window.postMessage = jest.fn();

    expect(window.postMessage).not.toHaveBeenCalled();
    const redirectButton = getByText('Click here');
    userEvent.click(redirectButton);

    expect(window.postMessage).toBeCalled();
  });
});
