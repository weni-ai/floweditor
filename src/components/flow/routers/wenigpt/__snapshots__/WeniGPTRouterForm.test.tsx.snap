// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeniGPTRouterForm should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Save",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="call_webhook"
  title="Call webhook"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "aliases": Array [
          "split_by_webhook",
        ],
        "component": [Function],
        "description": "Call a webhook",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Call webhook",
        "type": "call_webhook",
        "visibility": Array [
          "messaging",
          "messaging_background",
          "voice",
        ],
      }
    }
    onChange={[MockFunction]}
  />
  <div
    className="content"
  >
    <div
      className="knowledge_base"
    >
      <SelectElement
        entry={
          Object {
            "value": null,
          }
        }
        key="knowledge_base_select"
        name="Knowledge Base"
        nameKey="label"
        onChange={[Function]}
        options={
          Array [
            Object {
              "content": Object {
                "content": [Circular],
                "id": "1",
                "intelligence": "Intelligence 1",
                "name": "Base 1",
              },
              "id": "1",
              "intelligence": "Intelligence 1",
              "label": "Intelligence 1 - Base 1",
              "name": "Base 1",
            },
            Object {
              "content": Object {
                "content": [Circular],
                "id": "2",
                "intelligence": "Intelligence 1",
                "name": "Base 2",
              },
              "id": "2",
              "intelligence": "Intelligence 1",
              "label": "Intelligence 1 - Base 2",
              "name": "Base 2",
            },
            Object {
              "content": Object {
                "content": [Circular],
                "id": "3",
                "intelligence": "Intelligence 2",
                "name": "Base 1",
              },
              "id": "3",
              "intelligence": "Intelligence 2",
              "label": "Intelligence 2 - Base 1",
              "name": "Base 1",
            },
          ]
        }
        placeholder="Select a knowledge base"
        showLabel={true}
        valueKey="id"
      />
      <div
        className="message"
      >
        <span>
          Don't have any knowledge base for your AI yet?
        </span>
        <span
          className="link"
          onClick={[Function]}
        >
          Click here
        </span>
      </div>
    </div>
    <TextInputElement
      autocomplete={true}
      entry={
        Object {
          "value": "@input.text",
        }
      }
      name="Insert an expression to be used as input"
      onChange={[Function]}
      placeholder="Ex: @input.text"
      showLabel={true}
      textarea={true}
    />
  </div>
  <OptionalTextInput
    maxLength={64}
    name="Save as result"
    onChange={[Function]}
    toggleText="Save as"
    value={
      Object {
        "value": "Result",
      }
    }
  />
</Dialog>
`;

exports[`WeniGPTRouterForm updates should save changes 1`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "input": "Updated expression input data",
          "knowledge_base": "3",
          "result_name": "My Webhook Result",
          "type": "call_wenigpt",
          "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
        },
        Object {
          "destination_uuid": null,
          "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "Success",
            ],
            "category_uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
            "type": "has_only_text",
            "uuid": "6c22884b-0e35-4206-982e-18320691eda9",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
            "name": "Success",
            "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
          },
          Object {
            "exit_uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
            "name": "Failure",
            "uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
          },
        ],
        "default_category_uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
        "operand": "@results.my_webhook_result.category",
        "type": "switch",
      },
      "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_wenigpt",
    },
  },
]
`;
